EIPS/EIP-8.md

```diff
@@ -0,0 +1,148 @@
+network protocol upgrades (as long as backwards-compatibility is maintained).
+
+### Rationale
+
+#### Changes to the devp2p Wire Protocol
+
+All clients currently contain statements such as the following:
+
+```python
+# pydevp2p/p2p_protocol.py
+if data['version'] != proto.version:
+    log.debug('incompatible network protocols', peer=proto.peer,
+        expected=proto.version, received=data['version'])
+    return proto.send_disconnect(reason=reasons.incompatibel_p2p_version)
+```
+
+These checks make it impossible to change the version or structure of the initial packet.
+Dropping them enables switching to a newer protocol version: Clients implementing a newer
+version simply send a packet with higher version and possibly additional list elements.
+
+* If such a packet is received by a node with lower version, it will blindly assume that
+  the remote end is backwards-compatible and respond with the old handshake.
+* If the packet is received by a node with equal version, new features of the protocol can
+  be used.
+* If the packet is received by a node with higher version, it can enable
+  backwards-compatibility logic or drop the connection.
+
+#### Changes to the RLPx TCP Handshake
+
+Discussions of the RLPx v5 changes (chunked packets, change to key derivation) have
+faltered in part because the v4 handshake encoding provides only one in-band way to add a
+version number: shortening the random portion of the nonce. Even if the RLPx v5 handshake
+proposal were accepted, future upgrades are hard because the handshake packet is a fixed
+size ECIES ciphertext with known layout.
+
+I propose the following changes to the handshake packets:
+
+* Adding the length of the ciphertext as a plaintext header.
+* Encoding the body of the handshake as RLP.
+* Adding a version number to both packets in place of the token flag (unused).
+* Removing the hash of the ephemeral public key (it is redundant).
+
+These changes make it possible to upgrade the RLPx TCP transport protocol in the same
+manner as described for the other protocols, i.e. by adding list elements and bumping the
+version. Since this is the first change to the RLPx handshake packet, we can seize the
+opportunity to remove all currently unused fields.
+
+Additional data is permitted (and in fact required) after the RLP list because the
+handshake packet needs to grow in order to be distinguishable from the old format. Clients
+can employ logic such as the following pseudocode to handle both formats simultaneously:
+
+```go
+packet = read(307, connection)
+if decrypt(packet) {
+    // process as old format
+} else {
+    size = unpack_16bit_big_endian(packet)
+    packet += read(size-307, connection)
+    if !decrypt(packet) {
+        // error
+    }
+    // process as new format
+}
+```
+
+Implementations may choose to add a random amount of bytes to vary the size of the packet.
+Doing so helps prevent statistical traffic classification.
```

Comment:
@karalabe Yes – although I'm not a trained cryptanalyst like others whom I've worked with on rlpx. The ciphertext will always be at least N bytes, so over time, the fact that the value is greater than N and rest is random leaks information. If the value is cleartext it can also be reliably manipulated which can create other problems; flipping a few bits to increase the value would cause a longer read which results in timing out the connection.

<3 that I was able to sync up and send eth transactions from several locations where I couldn't use github.